### ISSUE TITLE
Making all error types explicits instead of inferred

### ISSUE THREAD

**Issue by Arwalk** [Arwalk/zig-protobuf#77]
I realize that in all our methods, we only use the inferred error type. This is overall not a problem, but when i see that #74 happened, this makes me think that it would be a nice improvement to explicitely set the errors in our method's interface.

Also, i feel like any serious error handling code would, for example, check if it's an invalid input or an allocation error, and recover if necessary.

the json codec would be optional on this subject, i'm not sure if it's easily feasible to keep it explicit.

**Comment by 42LM** [comment]
Hey there @Arwalk, i would like to help out here and this seems like a good task to tackle to read and get to know the code.

So if i understand correctly you want to have explicit error return like in:
https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L183
```zig
Whatever.Error!void
```

and not the inferred error type:
https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L34
```zig
!ManagedString
```

?

**Comment by Arwalk** [comment]
> Hey there [@Arwalk](https://github.com/Arwalk), i would like to help out here and this seems like a good task to tackle to read and get to know the code.
> 
> So if i understand correctly you want to have explicit error return like in:
> 
> [zig-protobuf/src/protobuf.zig](https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L183)
> 
> Line 183 in [42ccf14](/Arwalk/zig-protobuf/commit/42ccf142e5d6b7de886b766af7d287021c561dfd)
> 
>  fn insert_raw_varint(pb: *ArrayList(u8), size: u64, start_index: usize) Allocator.Error!void { 
> Whatever.Error!void
> and not the inferred error type:
> 
> [zig-protobuf/src/protobuf.zig](https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L34)
> 
> Line 34 in [42ccf14](/Arwalk/zig-protobuf/commit/42ccf142e5d6b7de886b766af7d287021c561dfd)
> 
>  pub fn copy(str: []const u8, allocator: Allocator) !ManagedString { 
> !ManagedString
> ?

Exactly. Thanks for your interest and enthusiasm. 

As said also, this is not something to do for the JSON perser/serializer for now.

**Comment by 42LM** [comment]
> Exactly. Thanks for your interest and enthusiasm.
> 
> As said also, this is not something to do for the JSON perser/serializer for now.

Yes you mean the `WireDecoderIterator` right?

I will find some time to start this week 🙌.

**Comment by Arwalk** [comment]
> Yes you mean the WireDecoderIterator right?

No, WireDecoderITerator is just an internal tool to move from tag to tag in a encoded message. I'm thinking about everything related  to [pb_json_encode](https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L1265) and [pb_json_decode](https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L1275). Everything related to these functions is not important to handle for now.

**Comment by 42LM** [comment]
> > Yes you mean the WireDecoderIterator right?
> 
> No, WireDecoderITerator is just an internal tool to move from tag to tag in a encoded message. I'm thinking about everything related to [pb_json_encode](https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L1265) and [pb_json_decode](https://github.com/Arwalk/zig-protobuf/blob/42ccf142e5d6b7de886b766af7d287021c561dfd/src/protobuf.zig#L1275). Everything related to these functions is not important to handle for now.

Okey! Thanks for the clarification 💪 :) i will make sure to take a good look before i start.
